File: python/1249.minimum_remove_valid_parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        s_list = list(s)
        stack = []

        for i, char in enumerate(s_list):
            if char == "(":
                stack.append(i)
            elif char == ")":
                if len(stack) > 0:
                    stack.pop()
                else:
                    s_list[i] = ""
        while stack:
            i = stack.pop()
            s_list[i] = ""

        return "".join(s_list)

File: cpp/1249.minimum_remove_valid_parentheses.cpp
#include <string>
#include <vector>

class Solution {
 public:
    std::string minRemoveToMakeValid(std::string s) {
        std::string result;
        std::vector<int> _stack;

        for (std::int64_t i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                _stack.push_back(i);
            } else if (s[i] == ')') {
                if (!_stack.empty()) {
                    _stack.pop_back();
                } else {
                    s[i] = '#';
                }
            }
        }

        while (!_stack.empty()) {
            int e = _stack.back();
            _stack.pop_back();
            s[e] = '#';
        }

        for (auto ch : s) {
            if (ch != '#') {
                result.push_back(ch);
            }
        }
        return result;
    }
};
